clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3675,14 +4062,260 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     removeFieldMemberExpr(FD, ArrayType);
     return true;
   }
+  using SyclKernelFieldHandler::enterArray;
+  using SyclKernelFieldHandler::enterStruct;
+  using SyclKernelFieldHandler::handleNonDecompStruct;
+  using SyclKernelFieldHandler::handlePointerType;
+  using SyclKernelFieldHandler::handleScalarType;
+  using SyclKernelFieldHandler::handleSyclSpecialType;
+  using SyclKernelFieldHandler::handleUnionType;
+  using SyclKernelFieldHandler::leaveArray;
+  using SyclKernelFieldHandler::leaveStruct;
+};
+
+class FreeFunctionKernelBodyCreator : public SyclKernelFieldHandler {
+  SyclKernelDeclCreator &DeclCreator;
+  llvm::SmallVector<Stmt *, 16> BodyStmts;
+  FunctionDecl *FreeFunc;
+  SourceLocation KernelCallerSrcLoc; // KernelCallerFunc source location.
+  llvm::SmallVector<Expr *, 8> ArgExprs;
+
+  // Creates a DeclRefExpr to the ParmVar that represents the current field.
+  Expr *createParamReferenceExpr() {
+    ParmVarDecl *KernelParameter =
+        DeclCreator.getParamVarDeclsForCurrentField()[0];
+
+    QualType ParamType = KernelParameter->getOriginalType();
```

Comment:
Please rename `ParamType` to `FreeFunctionParamType`. It is difficult to follow the logic here since we are now dealing with parameters in 2 different contexts as opposed to generated parameter vs kernel object fields like before.